Bluetooth: unittest: DLE procedure routines (#23758)
zephyrproject-rtos/zephyr: tests/bluetooth/controller/ctrl_ull_dletime/CMakeLists.txt

# SPDX-License-Identifier: Apache-2.0

cmake_minimum_required(VERSION 3.13.1)

#
# the following allows us to test the different PHY configurations
# by setting the appropriate flags in the yaml file
#
if(CONFIG_BT_CTLR_PHY)
	add_compile_definitions(CONFIG_BT_CTLR_PHY)
endif(CONFIG_BT_CTLR_PHY)

if(CONFIG_BT_CTLR_PHY_2M)
	add_compile_definitions(CONFIG_BT_CTLR_PHY_2M)
endif(CONFIG_BT_CTLR_PHY_2M)

if(CONFIG_BT_CTLR_PHY_CODED)
	add_compile_definitions(CONFIG_BT_CTLR_PHY_CODED)
endif(CONFIG_BT_CTLR_PHY_CODED)

if(CONFIG_BT_CTLR_DATA_LENGTH_MAX)
	add_compile_definitions(CONFIG_BT_CTLR_DATA_LENGTH_MAX=${CONFIG_BT_CTLR_DATA_LENGTH_MAX})
endif(CONFIG_BT_CTLR_DATA_LENGTH_MAX)

include_directories(
        $ENV{ZEPHYR_BASE}/tests/bluetooth/controller/mock_ctrl/include
        $ENV{ZEPHYR_BASE}/include/bluetooth
        $ENV{ZEPHYR_BASE}/subsys/bluetooth
        $ENV{ZEPHYR_BASE}/subsys/bluetooth/controller
        $ENV{ZEPHYR_BASE}/subsys/bluetooth/controller/util
        $ENV{ZEPHYR_BASE}/subsys/bluetooth/controller/include

>         $ENV{ZEPHYR_BASE}/subsys/bluetooth/controller/ll_sw

Reviewer: Note, after the introduction of Zephyr CMake package, user's are no longer required to set `ZEPHYR_BASE` in the environment.

But `find_package(Zephyr)` will define the `ZEPHYR_BASE` variable, so this code should be located after `find_package`, and updated to just use `${ZEPHYR_BASE}` instead.

Author: @tejlmand That makes sense, however this change results in the error that there is no board 'unit_testing'
The alternative would be to not use unit-testing, but that makes it difficult to test for the different configurations.

Author: Note also that the 'sanity_check' scripts which is used for running these tests require that ZEPHYR_BASE is set in the environment.

The complaint about no board 'unit_testing' goes away by defining ZEPHYR_UNITTEST in the CMake file, but then the target executable get set with the wrong sources, and the 'find_package(ZephyrUnittest)' fails due to the target already being defined

Author: Setting NO_BOILERPLATE will allow the suggestion of using find_package to work, but it does not seem more elegant to me:

set(NO_BOILERPLATE TRUE)
find_package(Zephyr)
set(NO_BOILERPLATE FALSE)

Reviewer: Thanks for looking into this, but let me try to explain a little better what I meant, along with a small example.
When I wrote that the use of `ZEPHYR_BASE` had to come after `find_package` I meant that you need to reorder the code.

Made an example for you here: https://github.com/kruithofa/zephyr/pull/2
feel free to squash in the changes and continue from there.

This also made me notice that the unittest boilerplate should be cleaned up, so that application boilerplate and unittest boiler plate are better aligned.

Reviewer: Note, the `include_directories` still needs fixing.
For the unit_testing target, you can simply use `set(INCLUDE include/bluetooth subsys/bluetooth ....)` without `ZEPHYR_BASE` as this will be prepended by the unittest boilerplate.

But that will not match the normal `find_package(Zephyr)` boilerplate.
Which brings me back to my previous comment:
> This also made me notice that the unittest boilerplate should be cleaned up, so that application boilerplate and unittest boiler plate are better aligned.

Author: Thanks for the example! 
Notes:
1. I could not find any documentation on the 'set(INCLUDE'. Is that preferred over 'include_directories'?
2. For finding the ZephyrUnittest we still rely on '$ENV(ZEPHYR_BASE)'.  Since I use the sanitycheck script for running unittests I haven't been able to test how this works if ZEPHYR_BASE is not defined.
3. I've sidestepped the discrepancy between Zephyr and ZephyrUnittest by only allowing this to run as a unittest.

Reviewer: > Thanks for the example!
> Notes:
> 
>     1. I could not find any documentation on the 'set(INCLUDE'. Is that preferred over 'include_directories'?
> 

Not necessarily, but you remove the need for knowing `ZEPHYR_BASE` as it will automatically be prepended by the unittest boilerplate code.
I haven't found any official docs on this (yet another thing to improve), but it is mentioned here:
https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/testsuite/unittest.cmake#L9-L11
and used here:
 https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/testsuite/unittest.cmake#L87-L89

>     2. For finding the ZephyrUnittest we still rely on '$ENV(ZEPHYR_BASE)'.  Since I use the sanitycheck script for running unittests I haven't been able to test how this works if ZEPHYR_BASE is not defined.
> 

For testing code like this, I use CMake directly, as well as not having `ZEPHYR_BASE` in my environment.
This should also work when not using sanitycheck.

>     3. I've sidestepped the discrepancy between Zephyr and ZephyrUnittest by only allowing this to run as a unittest.

If you judge that as an acceptable limitation, I have no objections.
It sure simplifies the code, but doesn't remove my wish for wanting to align the boilerplates in the future.

)

#
# we test functionality of ull_conn, which is included directly in the test file
# so we do not list it here
#
FILE(GLOB ll_sw_sources
	$ENV{ZEPHYR_BASE}/subsys/bluetooth/controller/util/mem.c
	$ENV{ZEPHYR_BASE}/subsys/bluetooth/controller/util/memq.c

> 	$ENV{ZEPHYR_BASE}/subsys/bluetooth/controller/ll_sw/ull_tx_queue.c

Reviewer: same comment regarding `ZEPHYR_BASE` as above.

)

FILE(GLOB mock_sources
	$ENV{ZEPHYR_BASE}/tests/bluetooth/controller/mock_ctrl/src/ll.c
	$ENV{ZEPHYR_BASE}/tests/bluetooth/controller/mock_ctrl/src/ull.c
	$ENV{ZEPHYR_BASE}/tests/bluetooth/controller/mock_ctrl/src/ull_master.c
	$ENV{ZEPHYR_BASE}/tests/bluetooth/controller/mock_ctrl/src/ticker.c

> 	$ENV{ZEPHYR_BASE}/tests/bluetooth/controller/mock_ctrl/src/mayfly.c

Reviewer: same comment regarding `ZEPHYR_BASE` as above.

# SPDX-License-Identifier: Apache-2.0

cmake_minimum_required(VERSION 3.13.1)

if (NOT BOARD STREQUAL unit_testing)

> 	message(FATAL_ERROR "This project can only be used for unit-testing.")

Reviewer: Just to help the user a bit here, so when he uses a wrong board, then please print the correct name with an underscore, and maybe a bit more to really make it clear.
```suggestion
  message(FATAL_ERROR "This project can only be used with `-DBOARD=unit_testing`")
```

# SPDX-License-Identifier: Apache-2.0

cmake_minimum_required(VERSION 3.13.1)

#
# the following allows us to test the different PHY configurations
# by setting the appropriate flags in the yaml file
#
if(CONFIG_BT_CTLR_PHY)

> add_definitions(-DCONFIG_BT_CTLR_PHY)

Author: Replaced add_definitions with add_compile_definitions

endif(CONFIG_BT_CTLR_PHY)

if(CONFIG_BT_CTLR_PHY_2M)
add_definitions(-DCONFIG_BT_CTLR_PHY_2M)
endif(CONFIG_BT_CTLR_PHY_2M)

if(CONFIG_BT_CTLR_PHY_CODED)
add_definitions(-DCONFIG_BT_CTLR_PHY_CODED)
endif(CONFIG_BT_CTLR_PHY_CODED)

if(CONFIG_BT_CTLR_DATA_LENGTH_MAX)
add_definitions(-DCONFIG_BT_CTLR_DATA_LENGTH_MAX=${CONFIG_BT_CTLR_DATA_LENGTH_MAX})

> set (CMAKE_C_FLAGS "-DCONFIG_BT_CTLR_DATA_LENGTH_MAX=${CONFIG_BT_CTLR_DATA_LENGTH_MAX}")

Reviewer: just a small note.
This will clear any existing `CMAKE_C_FLAGS`, but as it is in the main CMakeLists.txt file, it should be safe.
So this is more a general info about being careful when setting variables this way.

Author: agree, the set should be removed, since it  is the add_compile_definitions

endif(CONFIG_BT_CTLR_DATA_LENGTH_MAX)

> 

Reviewer 2: Unnecessary empty line (you should need just one)

Author: Extra empty lines removed

include_directories(
        $ENV{ZEPHYR_BASE}/tests/bluetooth/controller/mock_ctrl/include
        $ENV{ZEPHYR_BASE}/include/bluetooth
        $ENV{ZEPHYR_BASE}/subsys/bluetooth
        $ENV{ZEPHYR_BASE}/subsys/bluetooth/controller
        $ENV{ZEPHYR_BASE}/subsys/bluetooth/controller/util
        $ENV{ZEPHYR_BASE}/subsys/bluetooth/controller/include
        $ENV{ZEPHYR_BASE}/subsys/bluetooth/controller/ll_sw
)

#
# we test functionality of ull_conn, which is included directly in the test file
# so we do not list it here
#
FILE(GLOB ll_sw_sources
	$ENV{ZEPHYR_BASE}/subsys/bluetooth/controller/util/mem.c
	$ENV{ZEPHYR_BASE}/subsys/bluetooth/controller/util/memq.c
	$ENV{ZEPHYR_BASE}/subsys/bluetooth/controller/ll_sw/ull_tx_queue.c
)

FILE(GLOB mock_sources
	$ENV{ZEPHYR_BASE}/tests/bluetooth/controller/mock_ctrl/src/ll.c
	$ENV{ZEPHYR_BASE}/tests/bluetooth/controller/mock_ctrl/src/ull.c
	$ENV{ZEPHYR_BASE}/tests/bluetooth/controller/mock_ctrl/src/ull_master.c
	$ENV{ZEPHYR_BASE}/tests/bluetooth/controller/mock_ctrl/src/ticker.c
	$ENV{ZEPHYR_BASE}/tests/bluetooth/controller/mock_ctrl/src/mayfly.c
)

FILE(GLOB app_sources
        src/*.c
)

> if(BOARD STREQUAL unit_testing)

Reviewer 3: Please wait until https://github.com/zephyrproject-rtos/zephyr/pull/23875#pullrequestreview-383695647 is merged and then use `find_package()` instead.

Author: Done

  project(bluetooth_ctrl_ull_conn)
  list(APPEND SOURCES ${app_sources} ${ll_sw_sources} ${mock_sources})

>   include($ENV{ZEPHYR_BASE}/subsys/testsuite/unittest.cmake)

Author: Rebased and using find_package now

else()
#
# Note: not working
#
target_sources(app PRIVATE ${app_sources} ${ll_sw_sources} ${mock_sources})

> include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)

Author: Noted
